# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmGC.h
# Commit: 1a73cfde72e8
# Full Hash: 1a73cfde72e8f4d19bdcfdbffa9fbc78bba01e24
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2021-09-28 09:54:33
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Introduce indirect stubs to optimize call_indirect. r=lth
#   
#   This patch introduces indirect stubs.
#   An indirect stub is a stub that takes care of any switching activities needed for call_indirect.
#   Before this patch, we have to conservatively assume that any call_indirect's target can be from a foreign instance,
# ==============================================================================

diff -r da3c5d8492db -r 1a73cfde72e8 js/src/wasm/WasmGC.h
--- a/js/src/wasm/WasmGC.h	Tue Sep 28 05:16:23 2021 +0000
+++ b/js/src/wasm/WasmGC.h	Tue Sep 28 05:24:15 2021 +0000
@@ -72,6 +72,21 @@
   // is also noted.  This is used in Instance::traceFrame to check that the
   // TrapExitDummyValue is in the expected place in the frame.
 
+  // StackMap and indirect stubs.
+  // StackMaps track every word in wasm frame except indirect stubs words
+  // because these stubs can be dynamically added or not at runtime.
+  // For example, when we do `call_indirect foo` for some cross-instance call
+  // clang-format off
+  // |   some stack arg     | <- covered by StackMap.
+  // |   caller TLS slot    | <- covered by StackMap.
+  // |   callee TLS slot    | <- covered by StackMap.
+  // |   return pc: caller  | <- At runtime we read the tag below and skip this Frame in Instance::traceFrame.
+  // |   callerFP (tag)     |
+  // |   some stubs data    |
+  // |   return pc: to stub | <- callee's wasm::Frame, covered by StackMap.
+  // |   caller FP          |
+  // clang-format on
+
   // The total number of stack words covered by the map ..
   uint32_t numMappedWords : 30;
 